policy/query-builder/query-builder.go

```diff
@@ -0,0 +1,214 @@
+		"product":       "(product = LOWER(@%s))",
+		"producerUid":   "(producerUid IN (%s))",
+		"renewMonth":    "((isDeleted = false OR isDeleted IS NULL) AND (EXTRACT(MONTH FROM startDate) = CAST(@%s AS INTEGER)))",
+		"paid":          "((isDeleted = false OR isDeleted IS NULL) AND (isPay = true))",
+		"unpaid":        "((isDeleted = false OR isDeleted IS NULL) AND (isPay = false))",
+		"recurrent":     "((isDeleted = false OR isDeleted IS NULL) AND (hasMandate = true))",
+		"notRecurrent":  "((isDeleted = false OR isDeleted IS NULL) AND (hasMandate = false OR hasMandate IS NULL))",
+	}
+
+	orClausesKeys = []string{"status", "payment"}
+)
+
+type QueryBuilder interface {
+	BuildQuery(map[string]string) (string, map[string]interface{})
+}
+
+type BigQueryQueryBuilder struct {
+	tableName       string
+	tableAlias      string
+	randomGenerator func() string
+}
+
+func NewBigQueryQueryBuilder(tableName, tableAlias string, randomGenerator func() string) BigQueryQueryBuilder {
+	if randomGenerator == nil {
+		randomGenerator = func() string {
+			var (
+				letters  = []rune("abcdefghijklmnopqrstuvwxyz")
+				alphanum = []rune("123456789abcdefghijklmnopqrstuvwxyz")
+			)
+			rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+			s := make([]rune, 12)
+			s[0] = letters[rnd.Intn(len(letters))]
+			for i := range s[1:] {
+				s[i+1] = alphanum[rnd.Intn(len(alphanum))]
+			}
+			return string(s)
+		}
+	}
+	return BigQueryQueryBuilder{
+		tableName:       tableName,
+		tableAlias:      tableAlias,
+		randomGenerator: randomGenerator,
+	}
+}
+
+func (qb *BigQueryQueryBuilder) getAllowedParams(params map[string]string) []string {
+	paramsKeys := lib.GetMapKeys(params)
+	for _, value := range paramsHierarchy {
+		for k, v := range value {
+			if lib.SliceContains(paramsKeys, k) {
+				return v
+			}
+		}
+	}
+	return nil
+}
+
+func (qb *BigQueryQueryBuilder) filterParams(params map[string]string, allowedParams []string) map[string]string {
+	paramsKeys := lib.GetMapKeys(params)
+	for _, key := range paramsKeys {
+		if !lib.SliceContains(allowedParams, key) {
+			delete(params, key)
+		}
+	}
+	return params
+}
+
+func (qb *BigQueryQueryBuilder) processOrClauseParam(paramValue string) string {
+	whereClauses := make([]string, 0)
+	statusList := strings.Split(paramValue, ",")
+	for _, status := range statusList {
+		if val, ok := paramsWhereClause[status]; ok && val != "" {
+			whereClauses = append(whereClauses, val)
+		}
+	}
+	return "(" + strings.Join(whereClauses, " OR ") + ")"
+}
+
+func (qb *BigQueryQueryBuilder) processProducerUidParam(paramKey, paramValue string, queryParams map[string]interface{}) string {
+	tmp := make([]string, 0)
+	for _, uid := range strings.Split(paramValue, ",") {
+		randomIdentifier := qb.randomGenerator()
+		queryParams[randomIdentifier] = uid
+		tmp = append(tmp, fmt.Sprintf("'@%s'", randomIdentifier))
+	}
+	return fmt.Sprintf(paramsWhereClause[paramKey], strings.Join(tmp, ", "))
+}
```

Comment:
`paramKey` could be hardcoded since the function is used exclusively by `"producerUid"`